### DISCUSSION TITLE
Can you make chorded/modal/nested mappings?

### DISCUSSION THREAD

**Discussion by rawsp33d** [wezterm/wezterm#1812] [Q&A] [answered]
Can you make nested mappings such as `LEADER => t => n`, where you would first hit `LEADER`, then `t`, and then `n`?

I like making "modes" in my tools as much as possible. Vim, tmux, i3, and ideally wezterm if it would be so kind.

**Comment by wez** [comment, marked as answer]
Can you give some concrete examples of the sort of thing you'd want to configure in wezterm?

There's a lot of complexity around keyboard processing, both in terms of the OS specific parts and in terms of the configuration, so if we were to add something like this, it needs some careful thought around how the config looks and feels as well as how the feature actually gets implemented!

I'm familiar with vim's modes, but I've never gone crazy with configuring keybindings in vim itself.  I've not tried anything complex in tmux input configuration either, and I've never felt comfortable with i3, so: I just don't have a good sense of what sort of things you might configure or how that might look.

**Reply by rawsp33d** [reply]
Yeah of course. Let's say I want to open a new split to the right.

- In vim, I hit `<space>bl` which is: `<a prefix key>`, followed by `b` for "buffer", followed by `l` for "right" in vim
- In tmux, I hit `<ctrl-space>bl` which is: `<a prefix key>`, followed by `b` for "buffer" (just to stay consistent with vim... they aren't called buffers in tmux), followed by `l` for "right"

But those modes have other mappings as well. In vim for example:
- `h,j,k,l` all open panes in that direction
- `H,J,K,L` clones the pane in that direction
- `c` closes the pane (`:bdelete`)
  - `C` forces a close (`:bdelete!`)
- `r` reloads the pane (`:edit`)
  - `R` forces a reload (`:edit!`)
- `w` writes the pane (`:w`)
  - `W` forces a write (`:w!`)
- `s` saves the pane (`:update`)
  - `S` forces a save (`:update!`)

There are more, but you get the idea. Tmux has far fewer mappings in the buffer/pane "mode", but where possible it's consistent. `<prefix>bc` and `<prefix>bC` for example do the same thing in closing and force closing a pane.

But that's all one mode. Then I have a tabs mode in both, wherein you navigate, create, delete, reorder, and other operations on tabs. I have windows mode that operate on full windows. And far more modes especially in vim.

I have these same kind of modes in my window manager i3. In i3, modes are first-class features and actually very much embraced and recommended by the project which is one of the reasons I love it. I have an `APPS` mode for example that opens common apps like my media player: `<super+space>am`, as in `<prefix>`, `a` for "apps", `m` for media. In i3 specifically I have quite a few nested modes as well, where hitting some key in some mode might take you to a deeper level or another mode.

Another notable common theme across all of these is a "resize" mode. In vim, tmux, and i3 I have a resize mode that resizes panes as you press `h,j,k,l` but keeps you in the resize mode, so you're just hitting `h,h,h,h,h,h` (or `ctrl-h` or `shift-h,shift-h` or `alt-h` for bigger/smaller increments) and then `<esc>` to exit the mode.

I even have modes in my keyboard thanks to a project named [kmonad](https://github.com/kmonad/kmonad). I don't do nearly as much here, but I have a prefix key that if I hit `<prefix>q` it changes my keyboard at a hardware level to qwerty, `<prefix>d` for dvorak, `<prefix>g` to open up a custom layout I made for games where it's qwerty shifted one key to the right, so WASD is actually ESDF with my index finger on the F key nipple.

TL;DR: `<prefix><mode><mapping>`. More generally, just allowing keys to be bound to arbitrary named contexts, but allowing other keys to switch to those contexts.

**Reply by rawsp33d** [reply]
I think there are three main high-level ways to go about it:
1. vim and [sxhkd](https://github.com/baskerville/sxhkd) let you write mappings in a chorded way, so `abc` means `a`, then `b`, then `c`
  Thus there is no "mode" concept, no logical grouping. 
2. i3 lets you create actual first-class "modes" with a special syntax, and then allows you to map keys to switch to said modes.
  Since you can bind a key to switch to a mode, you can have those kinds of binds in modes too, and thus easily "nest" modes
3. tmux lets you bind keys scoped to what it calls "tables", which are essentially modes. And it lets you bind keys to switch to a specific table/mode context.

1 is probably the easiest to implement. 2 Is probably the hardest to implement and I don't think it would translate very well to wezwerm. 3 Is IMHO the best way to go about it for this project.

Tmux did not always have this feature, and I used to have a hacky workaround where hitting `<prefix>b` for example would run a shell script I wrote that would then make more, temporary tmux mappings. But by popular request they later added this concept into a project that was not built with it in mind, and it is surprisingly perfect. As a lot of people use wezterm in ways very similar to tmux, I think taking a leaf out of their book would be perfect. Also full disclosure, I would like to replace my everyday tmux usage entirely with wezterm, with this feature being the only one holding me back for now.

**Reply by wez** [reply]
I agree that something like 3 sounds interesting. Some users have expressed the desire to apply different sets of key bindings on a per-pane basis so this might help with that.

These issues are also related:

* https://github.com/wez/wezterm/issues/1129
* https://github.com/wez/wezterm/issues/993 (not 100%, but there's some overlap and it's worth consider together)

**Reply by rawsp33d** [reply]
Just to show how this was kind of added into tmux in a way that didn't drastically alter any syntax:

```tmux
# tmux.conf

# normal mapping
bind c kill-pane

# they added the ability to scope mappings with the `-T` argument ("T" for "key [T]able")
bind -T SOME_MODE c kill-pane
# and the `switch-client` command to switch to one of these tables
bind p switch-client -T SOME_MODE
```

Here `<prefix>pc` would kill a pane. I guess one elephant in this example's room is that all tmux mappings require a prefix unless you pass an argument to bind that makes it a global mapping. Wezterm doesn't have this implicit "prefix" concept which makes it even easier to understand and work with.

**Reply by wez** [reply]
https://github.com/wez/wezterm/commit/54d503de9d85350fdeae75f600767072d1c22470 allows:

```lua
local wezterm = require 'wezterm';

-- Show which table is active in the status area
wezterm.on("update-right-status", function(window, pane)
  local name = window:active_key_table()
  if name then
    name = "TABLE: " .. name
  end
  window:set_right_status(name or "")
end);

return {
  debug_key_events = true,
  keys = {
    -- Activate the "woot" table as a one-shot with
    -- a 2 second timeout, after which it will restore
    -- the default table.
    {
      key="a", mods="CTRL",
      action=wezterm.action{
        ActivateKeyTable={
          name="woot",
          timeout_milliseconds=2000,
        }
      }
    },

    -- Activate the "woot" table.
    -- The table will remain active until explicitly popped
    -- by the `PopKeyTable` action. See the Escape binding
    -- in the woot table below!
    {
      key="b", mods="CTRL",
      action=wezterm.action{
        ActivateKeyTable={
          name="woot",
          one_shot = false,
        }
      }
    },

    -- Activate the "woot" table as a one-shot with
    -- no timeout. It will remain active until a key is pressed,
    -- after which is will restore the default table.
    {
      key="c", mods="CTRL",
      action=wezterm.action{
        ActivateKeyTable={
          name="woot",
        }
      }
    },
  },

  key_tables = {
    woot = {
      {key="a", action=wezterm.action{SendString="woot"}},
      {key="Escape", action="PopKeyTable"},
    },
  },
}
```

**Reply by bew** [reply]
♥️♥️♥️♥️

**Reply by rawsp33d** [reply]
What the hell, man, I was just starting a discussion. And then you go and actually implement the thing within a couple hours?! We are unworthy. <3

**Reply by wez** [reply]
I've added docs for this stuff: https://wezfurlong.org/wezterm/config/key-tables.html

**Reply by MuhammedZakir** [reply]
> I've added docs for this stuff: https://wezfurlong.org/wezterm/config/key-tables.html

The leader key is assigned to `Ctrl+Shift+space`, but it conflicts with default key assignment of `QuickSelect`. I think `Ctrl+space` is better for this as one reason mentioned in the doc is to reduce the need to hold many keys at once.

---

In https://wezfurlong.org/wezterm/config/default-keys.html:
> `CTRL+SHIFT`	`(Space)`	`QuickSelect` (since: 20210502-130208-bff6815d)

Why is `space` inside brackets?

**Reply by bew** [reply]
> I think `Ctrl+space` is better for this as one reason mentioned in the doc is to reduce the need to hold many keys at once.

But `Ctrl+Space` is a valid terminal key, you'd be _stealing_ a key usable in programs `¯\_(ツ)_/¯`



> Why is `space` inside brackets?

Because there is actually a `` ` ` `` before, which is not very visible.
Maybe it'd be simpler to just remove it and only keep `` `Space` ``
